rados/operation.go

```diff
@@ -0,0 +1,127 @@
+package rados
+
+import "C"
+
+import (
+	"fmt"
+	"strings"
+)
+
+// The file operation.go exists to support both read op and write op types that
+// have some pretty common behaviors between them. In C/C++ its assumed that
+// the buffer types and other pointers will not be freed between passing them
+// to the action setup calls (things like rados_write_op_write or
+// rados_read_op_omap_get_vals2) and the call to Operate(...).  Since there's
+// nothing stopping one from sleeping for hours between these calls, or passing
+// the op to other functions and calling Operate there, we want a mechanism
+// that's (fairly) simple to understand and won't run afoul of Go's garbage
+// collection.  That's one reason the operation type tracks the steps (the
+// parts that track complex inputs and outputs) so that as long as the op
+// exists it will have a reference to the step, which will have references
+// to the C language types.
+
+type opKind string
+
+const (
+	readOp  opKind = "read"
+	writeOp opKind = "write"
+)
+
+// OperationError is an error type that may be returned by an Operate call.
+// It captures the error from the operate call itself and any errors from
+// steps that can return an error.
+type OperationError struct {
+	kind       opKind
+	OpError    error
+	StepErrors map[int]error
+}
+
+func (e OperationError) Error() string {
+	subErrors := []string{}
+	if e.OpError != nil {
+		subErrors = append(subErrors,
+			fmt.Sprintf("op=%s", e.OpError))
+	}
+	for idx, es := range e.StepErrors {
+		subErrors = append(subErrors,
+			fmt.Sprintf("Step#%d=%s", idx+1, es))
```

Comment:
why start counting with 1, and not 0?

Reply:
It seemed right at the time, but not I'm unsure. The goal is to be able to map back to operation steps that actually encountered an error. So, in the case that the entire error got printed/logged one could observe which operation failed. I think, at the time, I figured it would make more sense to use "human" numbering 1st step, 2nd step, etc. rather than "software" numbering. But now, I wonder how useful that'd be as it seems that requring the person reading the logs to know the exact order of the steps is a bit fishy.

Between this and @ansiwen other comments on this same type I am wondering if we ought to add ad name too. But we can do that later too. Opinions?

Reply:
Changed to use use zero indexing.